src/components/DocumentationTopic/PrimaryContent/DeclarationSource.vue

```diff
@@ -52,6 +51,111 @@ export default {
+    //
+    // This is implemented in a single pass loop where text tokens are updated
+    // at certain key points to insert spaces and newlines so that each
+    // parameter of a multi-parameter function gets its own line.
+    //
+    // Ideally this should be implemented with a tool like SwiftFormat in the
+    // future with the raw text before it gets tokenized for syntax
+    // highlighting—however, this post-tokenization JavaScript logic should be
+    // an improvement until that kind of work can be integrated.
+    //
+    // @param {Array} tokens The original syntax tokens.
+    //   See `DeclarationToken.props`
+    // @return {Array} A formatted version of the original tokens.
+    //   See `DeclarationToken.props`
+    formattedSwiftTokens: ({ tokens }) => {
+      const numSpacesForIndent = 4; // maybe this could be configurable in the future
+      const indent = ' '.repeat(numSpacesForIndent);
+      let indentedParams = false;
+      const newTokens = [];
+      let i = 0;
+      let j = 1;
+      let openParenTokenIndex = null;
+      let openParenCharIndex = null;
+      let closeParenTokenIndex = null;
+      let closeParenCharIndex = null;
+      let numUnclosedParens = 0;
+
+      // loop through every declaration token
+      while (i < tokens.length) {
+        // keep track of the current token and the next one (if any)
+        const token = tokens[i];
+        const newToken = { ...token };
+        const nextToken = j < tokens.length ? tokens[j] : undefined;
+
+        // loop through the token text to look for "(" and ")" characters
+        // eslint-disable-next-line no-plusplus
+        for (let k = 0; k < (token.text || '').length; k++) {
+          if (token.text.charAt(k) === '(') {
+            numUnclosedParens += 1;
+            // keep track of the token/character position of the first "("
+            if (openParenCharIndex == null) {
+              openParenCharIndex = k;
+              openParenTokenIndex = i;
+            }
+          }
+
+          if (token.text.charAt(k) === ')') {
+            numUnclosedParens -= 1;
+            // if this ")" balances out the number of "(" characters that have
+            // been seen, this is the one that pairs up with the first one
+            if (openParenTokenIndex !== null && numUnclosedParens === 0) {
+              closeParenCharIndex = k;
+              closeParenTokenIndex = i;
+            }
+          }
+        }
+
+        // if we find some text ending with ", " and the next token is the start
+        // of a new param, update this token text to replace the space with a
+        // newline followed by 4 spaces
+        if (token.text && token.text.endsWith(', ')
+          && nextToken && nextToken.kind === TokenKind.externalParam) {
+          newToken.text = `${token.text.trimEnd()}\n${indent}`;
+          indentedParams = true;
+        }
+
+        newTokens.push(newToken);
+        i += 1;
```

Comment:
Why do you have both `i` and `j` ? Is `j` just `i + 1`?

Reply:
Yes. `j` is just the next index (`i + 1`) so that we can peek at the next token to find the situation where one token is followed by another specific kind of token. I could remove the variable, but then we will have to repeat the `i + 1` expression in a couple of places—I don't mind making that change if it's confusing having so many variables.

Reply:
nah its fine, just wanted to make sure I understand. We can add a comment just for readability.